File: items.py
import assets
from config import *

class Item:

    credits_produced = 0

    def __init__(self, image, x, y):
        self.image = image
        self.name = ""
        self.description = ""
        self.x = x
        self.y = y
        self.tile = None
        self.cost = 0
        self.credits_produced = 0

    def render(self, surface):
        surface.blit(self.image, (self.x, self.y))


class FactoryItem(Item):

    def __init__(self, x, y):
        super().__init__(assets.factory_item, x, y)
        from tiles import FactoryTile
        self.tile = FactoryTile
        self.cost = 10
        self.name = "Factory"
        self.description = "Makes 3 pizzas. Requires connection to 1 Housing and 2 Farms."

class FarmItem(Item):

    credits_produced = 1

    def __init__(self, x, y):
        super().__init__(assets.farm_item, x, y)
        from tiles import Farm
        self.tile = Farm
        self.cost = 2
        self.credits_produced = 1
        self.name = "Farm"
        self.description = "Grows products to make pizza"

class RoadItem(Item):
    def __init__(self, x, y):
        super().__init__(assets.road_item, x, y)
        from tiles import RoadTile
        self.tile = RoadTile
        self.cost = 1
        self.name = "Road"
        self.description = "Connects tiles together"

class MonumentItem(Item):
    def __init__(self, x, y):
        super().__init__(assets.monument_item, x, y)
        from tiles import MonumentTile
        self.tile = MonumentTile
        self.cost = 20
        self.name = "Energy Drink Monument"
        self.description = "Red Bull is awesome."

class PizzeriaItem(Item):
    def __init__(self, x, y):
        super().__init__(assets.pizzeria_item, x, y)
        from tiles import PizzeriaTile
        self.tile = PizzeriaTile
        self.cost = 2
        self.name = "Pizzeria"
        self.description = "Makes 1 pizza. Requires connection to 1 Farm."

class HousingItem(Item):
    def __init__(self, x, y):
        super().__init__(assets.housing_item, x, y)
        from tiles import Housing
        self.tile = Housing
        self.cost = 3
        self.name = "Housing"
        self.description = "Supplies workers to Factories."

class DestructionItem(Item):
    def __init__(self, image, x, y):
        super().__init__(image, x, y)
        self.targets = ()

class AxeItem(DestructionItem):
    def __init__(self, x, y):
        super().__init__(assets.axe_item, x, y)
        self.cost = 1
        from tiles import ForestTile
        self.targets = (ForestTile, )
        self.name = "Axe"
        self.description = "Cuts down forests."



class Button:
    def __init__(self, image, x, y, engine):
        self.image = image
        self.x = x
        self.y = y
        self.width = ITEM_WIDTH
        self.height = ITEM_WIDTH
        self.engine = engine
        self.selected = False

    def render(self):
        self.engine.render(self.image, self.x, self.y)
        if(self.x <= self.engine.mouse_x <= self.x + self.width and self.y <= self.engine.mouse_y <= self.y + self.height):
            self.engine.render(assets.inventory_selector, self.x, self.y)

    def update(self):
        if(self.x <= self.engine.mouse_x <= self.x + self.width and self.y <= self.engine.mouse_y <= self.y + self.height):
            self.selected = True
        else:
            self.selected = False


    

File: items.js
import * as assets from './assets.js';
import { ITEM_WIDTH } from './config.js';
import {
    FactoryTile, Farm, RoadTile, MonumentTile, Tower, PizzeriaTile, Housing, ForestTile
} from './tiles.js';

export class Item {
    constructor(image, x, y) {
        this.image = image;
        this.name = '';
        this.description = '';
        this.x = x;
        this.y = y;
        this.tile = null;
        this.cost = 0;
        this.credits_produced = 0;
    }

    render(ctx) {
        ctx.drawImage(this.image, this.x, this.y);
    }
}

export class FactoryItem extends Item {
    constructor(x, y) {
        super(assets.factory_item, x, y);
        this.tile = FactoryTile;
        this.cost = 10;
        this.name = 'Factory';
        this.description = 'Makes 3 pizzas. Requires connection to 1 Housing and 2 Farms.';
    }
}

export class FarmItem extends Item {
    constructor(x, y) {
        super(assets.farm_item, x, y);
        this.tile = Farm;
        this.cost = 2;
        this.credits_produced = 1;
        this.name = 'Farm';
        this.description = 'Grows products to make pizza';
    }
}

export class RoadItem extends Item {
    constructor(x, y) {
        super(assets.road_item, x, y);
        this.tile = RoadTile;
        this.cost = 1;
        this.name = 'Road';
        this.description = 'Connects tiles together';
    }
}

export class MonumentItem extends Item {
    constructor(x, y) {
        super(assets.monument_item, x, y);
        this.tile = MonumentTile;
        this.cost = 20;
        this.name = 'Energy Drink Monument';
        this.description = 'Red Bull is awesome.';
    }
}

export class TowerItem extends Item {
    constructor(x, y) {
        super(assets.tower_item, x, y);
        this.tile = Tower;
        this.cost = 10;
        this.name = 'Military Tower';
        this.description = 'Defends from Nolats within 3 blocks';
    }
}

export class PizzeriaItem extends Item {
    constructor(x, y) {
        super(assets.pizzeria_item, x, y);
        this.tile = PizzeriaTile;
        this.cost = 2;
        this.name = 'Pizzeria';
        this.description = 'Makes 1 pizza. Requires connection to 1 Farm.';
    }
}

export class HousingItem extends Item {
    constructor(x, y) {
        super(assets.housing_item, x, y);
        this.tile = Housing;
        this.cost = 3;
        this.name = 'Housing';
        this.description = 'Supplies workers to Factories.';
    }
}

export class DestructionItem extends Item {
    constructor(image, x, y) {
        super(image, x, y);
        this.targets = [];
    }
}

export class AxeItem extends DestructionItem {
    constructor(x, y) {
        super(assets.axe_item, x, y);
        this.cost = 1;
        this.targets = [ForestTile];
        this.name = 'Axe';
        this.description = 'Cuts down forests.';
    }
}

export class Button {
    constructor(image, x, y, engine) {
        this.image = image;
        this.x = x;
        this.y = y;
        this.width = ITEM_WIDTH;
        this.height = ITEM_WIDTH;
        this.engine = engine;
        this.selected = false;
    }

    render() {
        this.engine.render(this.image, this.x, this.y);
        if (this.x <= this.engine.mouse_x && this.engine.mouse_x <= this.x + this.width
            && this.y <= this.engine.mouse_y && this.engine.mouse_y <= this.y + this.height) {
            this.engine.render(assets.inventory_selector, this.x, this.y);
        }
    }

    update() {
        if (this.x <= this.engine.mouse_x && this.engine.mouse_x <= this.x + this.width
            && this.y <= this.engine.mouse_y && this.engine.mouse_y <= this.y + this.height) {
            this.selected = true;
        } else {
            this.selected = false;
        }
    }
}
